### PR TITLE
Some configurable parameters

### PR DESCRIPTION
**Pull request by safanaj**
To use VPA on AWS EKS I needed to change the port for the admission controller webhook, so I would propose to have a flag for it.

For the recommender to be able to get historical metrics from prometheus I need to tune the metrics and labels to query prometheus properly, so I would add a bunch of flags for that.

This PR is just to share the ideas.

### CODE REVIEW

File: vertical-pod-autoscaler/pkg/recommender/input/history/history_provider.go [hunk 1]
```diff
@@ -26,10 +26,22 @@ import (
 	"k8s.io/autoscaler/vertical-pod-autoscaler/pkg/recommender/model"
 )
```
Comment (Rajat-0) [inline vertical-pod-autoscaler/pkg/recommender/input/history/history_provider.go]: I have a doubt. Why struct  "PrometheusHistoryProviderConfig"  and "prometheusHistoryProviderConfig" have almost similar name and similar fields.

File: vertical-pod-autoscaler/pkg/recommender/input/history/history_provider.go [hunk 2]
```diff
@@ -26,10 +26,22 @@ import (
...
 
-const (
-	historyLength  = "8d"
-	podLabelPrefix = "pod_label_"
-)
+type PrometheusHistoryProviderConfig struct {
```
Comment (Rajat-0) [inline vertical-pod-autoscaler/pkg/recommender/input/history/history_provider.go]: I have a doubt. Why struct "PrometheusHistoryProviderConfig" and "prometheusHistoryProviderConfig" have almost similar name and similar fields.

**Comment by bskiba** [comment]
I'll take a look tomorrow, sorry to keep you waiting.

File: vertical-pod-autoscaler/pkg/recommender/input/history/history_provider.go (hunk 2 above)
Comment (safanaj) [reply vertical-pod-autoscaler/pkg/recommender/input/history/history_provider.go]: yes, those are almost the same, and maybe the unexported one could be removed.

**Comment by safanaj** [comment]
Reading this issue https://github.com/kubernetes/autoscaler/issues/1551#issuecomment-456386360 , maybe could worth to add also flags to choose the metric names for: CPU and memory defaulting them as `container_cpu_usage_seconds_total` and `container_memory_usage_bytes` , but allow users to choose what could be more appropriate like `container_memory_working_set_bytes`

File: vertical-pod-autoscaler/pkg/admission-controller/main.go [hunk 3]
```diff
@@ -41,6 +42,7 @@ var (
 		tlsPrivateKey: flag.String("tls-private-key", "/etc/tls-certs/serverKey.pem", "Path to server certificate key PEM file."),
 	}
 
+	port      = flag.Int("port", 8000, "The port to linsten on.")
```
Comment (bskiba) [inline vertical-pod-autoscaler/pkg/admission-controller/main.go]: s/linsten/listen/

File: vertical-pod-autoscaler/pkg/recommender/input/history/history_provider.go [hunk 4]
```diff
@@ -56,25 +68,39 @@ type prometheusHistoryProvider struct {
 	prometheusClient PrometheusClient
 }
 
+var singletonHistoryProvider *prometheusHistoryProvider
```
Comment (bskiba) [inline vertical-pod-autoscaler/pkg/recommender/input/history/history_provider.go]: Why this change?

File: vertical-pod-autoscaler/pkg/recommender/input/history/history_provider.go [hunk 5]
```diff
@@ -56,25 +68,39 @@ type prometheusHistoryProvider struct {
...
+		ctrNameLabel:           config.CtrNameLabel,
+		cadvisorMetricsJobName: config.CadvisorMetricsJobName,
+	}
+	singletonHistoryProvider = &prometheusHistoryProvider{
+		prometheusClient: NewPrometheusClient(&http.Client{}, config.Address),
+	}
+	return singletonHistoryProvider
 }
 
 func getContainerIDFromLabels(labels map[string]string) (*model.ContainerID, error) {
```
Comment (bskiba) [inline vertical-pod-autoscaler/pkg/recommender/input/history/history_provider.go]: If you make config part of prometheusHistoryProvider, you can make these functions methods of *prometheusHistoryProvider. For example here 
`func(p *prometheusHistoryProvider) getContainerIDFromLabels(labels map[string]string) (*model.ContainerID, error) {`

and then use `p.config.sth`

**Comment by bskiba** [comment]
Golint is unhappy:
!!! 'golint' problems: 
/home/travis/gopath/src/k8s.io/autoscaler/vertical-pod-autoscaler/pkg/recommender/input/history/history_provider.go:29:6: exported type PrometheusHistoryProviderConfig should have comment or be unexported
FAILED

**Comment by bskiba** [comment]
Some initial comments. I am missing changes to tests.

I also see that a merge conflict appeared in the meantime :(

**Comment by bskiba** [comment]
I might have some more comments to the flag naming but let's get through the first comments for now.

File: vertical-pod-autoscaler/pkg/recommender/input/history/history_provider.go (hunk 2 above)
Comment (safanaj) [reply vertical-pod-autoscaler/pkg/recommender/input/history/history_provider.go]: removed the unexported one

File: vertical-pod-autoscaler/pkg/recommender/input/history/history_provider.go (hunk 5 above)
Comment (safanaj) [reply vertical-pod-autoscaler/pkg/recommender/input/history/history_provider.go]: done

**Comment by safanaj** [comment]
Sorry for the delayed response. First round of requested changes done.

**Comment by mwielgus** [comment]
The tests are failing
`
?   	k8s.io/autoscaler/vertical-pod-autoscaler/pkg/recommender	[no test files]
# k8s.io/autoscaler/vertical-pod-autoscaler/pkg/recommender/input/history [k8s.io/autoscaler/vertical-pod-autoscaler/pkg/recommender/input/history.test]
pkg/recommender/input/history/history_provider_test.go:51:3: unknown field 'prometheusJobName' in struct literal of type prometheusHistoryProvider
pkg/recommender/input/history/history_provider_test.go:64:3: unknown field 'prometheusJobName' in struct literal of type prometheusHistoryProvider
pkg/recommender/input/history/history_provider_test.go:75:3: unknown field 'prometheusJobName' in struct literal of type prometheusHistoryProvider
pkg/recommender/input/history/history_provider_test.go:102:3: unknown field 'prometheusJobName' in struct literal of type prometheusHistoryProvider
pkg/recommender/input/history/history_provider_test.go:129:3: unknown field 'prometheusJobName' in struct literal of type prometheusHistoryProvider
?   	k8s.io/autoscaler/vertical-pod-autoscaler/pkg/recommender/input	[no test files]
`

**Comment by bskiba** [comment]
I am still working on this review, ETA Monday

**Comment by bskiba** [comment]
@safanaj This looks good now, thanks you! Can you take a look at the failing test? Otherwise LGTM

**Comment by safanaj** [comment]
I fixed the tests to use the new prom history provider config struct

**Comment by bskiba** [comment]
Thanks!
/lgtm
/approve
